File: src/main/java/org/yah/tests/perceptron/Matrix.java
package org.yah.tests.perceptron;

public class Matrix {
    public static double[][] matrix(int rows, int cols) {
        return new double[rows][cols];
    }

    public static double[][] transpose(double[][] m) {
        double[][] t = matrix(m[0].length, m.length);
        for (int r = 0; r < m.length; r++) {
            for (int c = 0; c < m[r].length; c++) {
                t[c][r] = m[r][c];
            }
        }
        return t;
    }

    public static int maxRowIndex(double[][] m, int col) {
        int res = -1;
        double max = Double.MIN_VALUE;
        for (int row = 0; row < m.length; row++) {
            double v = m[row][col];
            if (v > max) {
                max = v;
                res = row;
            }
        }
        return res;
    }

    public static void zero(double[][] m) {
        for (int r = 0; r < m.length; r++) {
            for (int c = 0; c < m[r].length; c++) {
                m[r][c] = 0f;
            }
        }
    }


}

File: cpp/Matrix.cpp
#include "Matrix.h"
#include <algorithm>

void Matrix::create(int _rows, int _columns) {
	rows = _rows;
	columns = maxColumns = _columns;
	data = new double[(size_t)_columns * _rows];
	managed = true;
	zero();
}

void Matrix::set(int _rows, int _columns, double* _data)
{
	rows = _rows;
	maxColumns = columns = _columns;
	data = _data;
	offset = 0;
	managed = false;
}

int Matrix::slide(int _offset, int _columns) {
	offset = _offset;
	columns = std::min(_columns, maxColumns - offset);
	return columns;
}

void Matrix::zero() {
	memset(data + offset, 0, sizeof(double) * columns * rows);
}

void Matrix::sub(Matrix& m) {
	for (int c = 0; c < columns; c++)
	{
		double* col = column(c);
		double* mcol = m.column(c);
		for (int r = 0; r < rows; r++)
		{
			col[r] -= mcol[r];
		}
	}
}

void Matrix::mul(double s) {
	for (int c = 0; c < columns; c++)
	{
		double* col = column(c);
		for (int r = 0; r < rows; r++)
		{
			col[r] *= s;
		}
	}
}

int Matrix::maxRowIndex(int c) {
	int res = 0;
	double* col = column(c);
	for (int r = 1; r < rows; r++) {
		if (col[r] > col[res])
			res = r;
	}
	return res;
}

void Matrix::free() {
	rows = columns = maxColumns = 0;
	if (managed) {
		delete[]data;
		managed = false;
	}
	data = 0;
}